pyfr/plugins/turbulence.py

```diff
@@ -0,0 +1,257 @@
+import math
+import numpy as np
+
+from collections import defaultdict
+from pyfr.plugins.base import BaseSolverPlugin
+from pyfr.regions import BoxRegion
+from rtree.index import Index, Property
+
+def lcg(seed):
+    state = np.uint32(seed)
+    while True:
+        yield state
+        with np.errstate(over='ignore'):
+            state = state*np.uint32(747796405) + np.uint32(2891336453)
+
+def pcg32rxs_m_xs_random(state):
+    with np.errstate(over='ignore'):
+        rshift = np.uint32(state >> np.uint32(28))
+        state ^= state >> (np.uint32(4) + rshift)
+        state *= np.uint32(277803737)
+        state ^= state >> np.uint32(22)
+        return (state >> 8) * 5.9604644775390625e-8
+
+def binary_search(f):
+    lbnd = 0
+    ubnd = 1
+
+    while not f(ubnd):
+        lbnd = ubnd
+        ubnd *= 2
+
+    while ubnd - lbnd > 1:
+        mpnt = (lbnd + ubnd) // 2
+        if f(mpnt):
+            ubnd = mpnt
+        else:
+            lbnd = mpnt
+
+    return ubnd
+
+class TurbulencePlugin(BaseSolverPlugin):
+    name = 'turbulence'
+    systems = [ 'ac-navier-stokes', 'navier-stokes']
+    formulations = ['dual', 'std']
+    dimensions = [3]
+
+    def __init__(self, intg, cfgsect):
+        super().__init__(intg, cfgsect)
+
+        self.fdptype = fdptype = intg.backend.fpdtype
+
+        ac = intg.system.name.startswith('ac')
+
+        self.tstart = intg.tstart
+        self.tbegin = intg.tcurr
+        self.tnext = intg.tcurr
+        self.tend = intg.tend
+
+        self.seed = self.cfg.getint(cfgsect, 'seed')
+
+        self.eventdtype = [('tinit', fdptype), ('state', np.uint32),
+                           ('ts', fdptype), ('te', fdptype)]
+
+        gamma = self.cfg.getfloat('constants', 'gamma')
+        avgrho = self.cfg.getfloat(cfgsect, 'avg-rho')
+        avgmach = self.cfg.getfloat(cfgsect, 'avg-mach')
+        ti = self.cfg.getfloat(cfgsect, 'turbulence-intensity')
+        sigma = self.cfg.getfloat(cfgsect, 'sigma')
+        centre = self.cfg.getliteral(cfgsect, 'centre')
+        rax = np.array(self.cfg.getliteral(cfgsect, 'rot-axis'))
+        ran = np.radians(self.cfg.getfloat(cfgsect,'rot-angle'))
+        self.avgu = avgu = self.cfg.getfloat(cfgsect, 'avg-u')
+        self.ls = ls = self.cfg.getfloat(cfgsect, 'turbulence-length-scale')
+
+        gc = (2*sigma/(math.erf(1/sigma)*math.pi**0.5))**0.5
+        beta1 = -0.5/(sigma*ls)**2
+        beta2 = avgrho*(gamma - 1)*avgmach**2
+        beta3 = 0.01*ti*avgu*(gc/sigma)**3
+
+        self.ydim = ydim = self.cfg.getfloat(cfgsect, 'y-dim')
+        self.zdim = zdim = self.cfg.getfloat(cfgsect, 'z-dim')
+        
+        self.ymin = ymin = -ydim/2
+        self.zmin = zmin = -zdim/2
+
+        self.nvorts = int(ydim*zdim/(4*ls**2))
```

Comment:
This will round down, is this what we want?

Reply:
I think thats fine. I don't think it was actually specified in the paper.